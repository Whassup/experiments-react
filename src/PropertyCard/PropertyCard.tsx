// Exhibit A: Fat dependencies

import { FunctionComponent } from "react";

interface PropertyAttributesProps {
  beds: number;
  baths: number;
  carSpaces: number;
}

export const PropertyAttributes: FunctionComponent<PropertyAttributesProps> = ({
  beds,
  baths,
  carSpaces,
}) => (
  <div>
    <h2>Property Attributes</h2>
    <dl>
      <dt>Beds</dt>
      <dd aria-label="beds">{beds}</dd>

      <dt>Baths</dt>
      <dd aria-label="baths">{baths}</dd>

      <dt>Carspaces</dt>
      <dd aria-label="carSpaces">{carSpaces}</dd>
    </dl>
  </div>
);

export interface PropertyCardProps {
  title: string;
  href: string;
  attributes?: React.ReactNode;
  onSaveToggle: () => void;
  saved?: boolean;
}

export const PropertyCard: FunctionComponent<PropertyCardProps> = ({
  title,
  attributes,
  href,
  saved,
  onSaveToggle,
} ) => {
  return (
    <article>
      <h1>
        <a href={href}>{title}</a>
      </h1>
      {attributes}
      <a href={href}>View details page</a>
      <button type="button" onClick={onSaveToggle}>
        {saved ? "Remove" : " Save"}
      </button>
    </article>
  );
};

PropertyCard.defaultProps = {
  saved: false
}
