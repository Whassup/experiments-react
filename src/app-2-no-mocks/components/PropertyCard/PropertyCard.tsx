// Exhibit A: Fat dependencies

import { FunctionComponent } from "react";

export interface PropertyCardProps {
  title: string;
  href: string;
  attributes?: React.ReactNode;
  onSaveToggle: () => void;
  saved?: boolean;
}

export const PropertyCard: FunctionComponent<PropertyCardProps> = ({
  title,
  attributes,
  href,
  saved = false,
  onSaveToggle,
}) => {
  return (
    <article>
      <h1>
        <a href={href}>{title}</a>
      </h1>
      {attributes}
      <a href={href}>View details page</a>
      <button type="button" onClick={onSaveToggle}>
        {saved ? "Remove" : " Save"}
      </button>
    </article>
  );
};
