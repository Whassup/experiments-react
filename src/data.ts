// TODO:  Add a property image to this example
// Add a results list with pageinfo

export interface PropertyData {
    property : {
        id: string,
        address: {
            shortAddressDisplay: string,
        },
        attributes: {
            beds: number,
            baths: number,
            carSpaces: number
        }
    },
    propertyDetailsPage: {
        slug: string
    }
}

export const createPropertyData = (): PropertyData =>({
    property : {
        id: String(Math.floor(Math.random() * 1000)),
        address: {
            shortAddressDisplay: "7 Westview Court, Springvale South",
        },
        attributes: {
            beds: 1,
            baths: 2,
            carSpaces: 3
        }
    },
    propertyDetailsPage: {
        slug: "details/7-westview-court-springvale-south-3172-vic"
    }
});


export interface PropertyResultsData {
    results: PropertyData[];
    pageInfo: {
        totalResultsCount: number;
        hasMore: boolean;
    }
}

export const createPropertyResultsData = (): PropertyResultsData =>({
    results: [createPropertyData(), createPropertyData(), createPropertyData()],
    pageInfo: {
        totalResultsCount: 100,
        hasMore: true
    }
});