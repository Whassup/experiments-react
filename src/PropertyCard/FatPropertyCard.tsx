// Exhibit A: Fat dependencies

import { FunctionComponent } from "react";
import { data } from "./data";


interface PropertyCardProps {
    data: typeof data;
}

export const FatPropertyCard: FunctionComponent<PropertyCardProps> = ({ data }) => {
    const { property, propertyDetailsPage } = data;
    const { address: { shortAddressDisplay }, attributes: { beds, baths, carSpaces } } = property;
    const { slug } = propertyDetailsPage;

    return <article>
        <h1><a href={slug}>{shortAddressDisplay}</a></h1>
        <div>
            <h2>Property Attributes</h2>
            <dl>
                <dt>Beds</dt>
                <dd aria-label="beds">{beds}</dd>

                <dt>Baths</dt>
                <dd aria-label="baths">{baths}</dd>

                <dt>Carspaces</dt>
                <dd aria-label="carSpaces">{carSpaces}</dd>
            </dl>
        </div>
        <a href={slug}>View details page</a>
    </article>
}


export default function FatPropertyCardWithData () {
    return <FatPropertyCard data={data} />
}