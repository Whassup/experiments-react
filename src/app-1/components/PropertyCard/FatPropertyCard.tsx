// Exhibit A: Fat dependencies

import { FunctionComponent } from "react";
import { PropertyData } from "../../../data";


interface FatPropertyCardProps {
    data: PropertyData;
    onSaveToggle: (propertyData: PropertyData) => void;
    saved?: boolean;
}

export const FatPropertyCard: FunctionComponent<FatPropertyCardProps> = ({ data, saved, onSaveToggle}) => {
    const { property, propertyDetailsPage } = data;
    const { address: { shortAddressDisplay }, attributes: { beds, baths, carSpaces } } = property;
    const { slug } = propertyDetailsPage;

    return <article data-testid={`card-${property.id}${saved ? "--saved" : ""}`} >
        <h1><a href={slug}>{shortAddressDisplay}</a></h1>
        <div>
            <h2>Property Attributes</h2>
            <dl>
                <dt>Beds</dt>
                <dd aria-label="beds">{beds}</dd>

                <dt>Baths</dt>
                <dd aria-label="baths">{baths}</dd>

                <dt>Carspaces</dt>
                <dd aria-label="carSpaces">{carSpaces}</dd>
            </dl>
        </div>
        <a href={slug}>View details page</a>
        <button data-testid={`save-${property.id}`} type="button" onClick={() => onSaveToggle(data)}>{  saved ? "Remove" : "Save" }</button>
    </article>
}

FatPropertyCard.defaultProps = {
    saved : false
}