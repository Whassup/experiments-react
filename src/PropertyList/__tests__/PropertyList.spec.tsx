import { render, screen } from "@testing-library/react";
import { createPropertyResultsData } from "../../PropertyCard/data";
import {
  PropertyList,
  PropertyListNoResults,
  PropertyListView,
  PropertyListWithData
} from "../PropertyList";

const propertiesData = createPropertyResultsData();

describe("<PropertyListWithData />", () => {
  it("SHOULD render with mapped data", () => {
    const { asFragment } = render(
      <PropertyListWithData propertiesData={propertiesData} />
    );
    expect(asFragment()).toMatchSnapshot();
  });
});

describe("<PropertyListView />", () => {
  const noResultsContent = "no results";
  const resultsContent = "display the results";

  describe("GIVEN results", () => {
    beforeEach(() => {
      render(
        <PropertyListView
          hasResults={true}
          list={<>{resultsContent}</>}
          noResults={<>{noResultsContent}</>}
        />
      );
    });
    it("SHOULD display list", () => {
      const list = screen.getByText(resultsContent);
      expect(list).toBeInTheDocument();
    });

    it("SHOULD NOT display no results", () => {
      const noResults = screen.queryByText(noResultsContent);
      expect(noResults).not.toBeInTheDocument();
    });
  });

  describe("GIVEN no results", () => {
    beforeEach(() => {
      render(
        <PropertyListView
          hasResults={false}
          list={<>{resultsContent}</>}
          noResults={<>{noResultsContent}</>}
        />
      );
    });
    it("SHOULD NOT display list", () => {
      const list = screen.queryByText(resultsContent);
      expect(list).not.toBeInTheDocument();
    });
    it("SHOULD display no results", () => {
      const noResults = screen.getByText(noResultsContent);
      expect(noResults).toBeInTheDocument();
    });
  });
});

describe("<PropertyList />", () => {
  const propertyCardsContent = "PropertyCards";
  const savedPropertyCardsContent = "SavedPropertyCards";

  beforeEach(() => {
    render(
      <PropertyList
        propertyCards={<>{propertyCardsContent}</>}
        savedPropertyCards={<>{savedPropertyCardsContent}</>}
        totalCount={100}
        pageCount={10}
      />
    );
  });

  it("SHOULD display propertyCards", () => {
    const propertyCards = screen.getByText(propertyCardsContent);
    expect(propertyCards).toBeInTheDocument();
  });

  it("SHOULD display result summary", () => {
    const resultSummary = screen.getByRole("heading", {
      name: "Displaying 10 of 100 total results.",
    });
    expect(resultSummary).toBeInTheDocument();
  });

  it("SHOULD display savedPropertyCards", () => {
    const propertyCards = screen.getByText(savedPropertyCardsContent);
    expect(propertyCards).toBeInTheDocument();
  });
});

describe("<PropertyListNoResults />", () => {
  beforeEach(() => {
    render(<PropertyListNoResults />);
  });

  it("SHOULD render", () => {
    const { asFragment } = render(<PropertyListNoResults />);
    expect(asFragment()).toMatchSnapshot();
  });
});
