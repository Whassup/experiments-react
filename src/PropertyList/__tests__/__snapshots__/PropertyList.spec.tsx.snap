// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<PropertyListNoResults /> SHOULD render 1`] = `
<DocumentFragment>
  <section>
    <h1>
      Properties
    </h1>
    <p>
      No results found please try another search.
    </p>
  </section>
</DocumentFragment>
`;

exports[`<PropertyListWithData /> SHOULD render with mapped data 1`] = `
<DocumentFragment>
  <section>
    <h1>
      Properties
    </h1>
    <h2>
      Displaying 
      <strong>
        3
      </strong>
       of 
      <strong>
        100
      </strong>
       total results.
    </h2>
    <section>
      <article>
        <h1>
          <a
            href="details/7-westview-court-springvale-south-3172-vic"
          >
            7 Westview Court, Springvale South
          </a>
        </h1>
        <div>
          <h2>
            Property Attributes
          </h2>
          <dl>
            <dt>
              Beds
            </dt>
            <dd
              aria-label="beds"
            >
              1
            </dd>
            <dt>
              Baths
            </dt>
            <dd
              aria-label="baths"
            >
              2
            </dd>
            <dt>
              Carspaces
            </dt>
            <dd
              aria-label="carSpaces"
            >
              3
            </dd>
          </dl>
        </div>
        <a
          href="details/7-westview-court-springvale-south-3172-vic"
        >
          View details page
        </a>
        <button
          type="button"
        >
           Save
        </button>
      </article>
      <article>
        <h1>
          <a
            href="details/7-westview-court-springvale-south-3172-vic"
          >
            7 Westview Court, Springvale South
          </a>
        </h1>
        <div>
          <h2>
            Property Attributes
          </h2>
          <dl>
            <dt>
              Beds
            </dt>
            <dd
              aria-label="beds"
            >
              1
            </dd>
            <dt>
              Baths
            </dt>
            <dd
              aria-label="baths"
            >
              2
            </dd>
            <dt>
              Carspaces
            </dt>
            <dd
              aria-label="carSpaces"
            >
              3
            </dd>
          </dl>
        </div>
        <a
          href="details/7-westview-court-springvale-south-3172-vic"
        >
          View details page
        </a>
        <button
          type="button"
        >
           Save
        </button>
      </article>
      <article>
        <h1>
          <a
            href="details/7-westview-court-springvale-south-3172-vic"
          >
            7 Westview Court, Springvale South
          </a>
        </h1>
        <div>
          <h2>
            Property Attributes
          </h2>
          <dl>
            <dt>
              Beds
            </dt>
            <dd
              aria-label="beds"
            >
              1
            </dd>
            <dt>
              Baths
            </dt>
            <dd
              aria-label="baths"
            >
              2
            </dd>
            <dt>
              Carspaces
            </dt>
            <dd
              aria-label="carSpaces"
            >
              3
            </dd>
          </dl>
        </div>
        <a
          href="details/7-westview-court-springvale-south-3172-vic"
        >
          View details page
        </a>
        <button
          type="button"
        >
           Save
        </button>
      </article>
    </section>
    <h1>
      Saved Properties
    </h1>
    <section />
  </section>
</DocumentFragment>
`;
